AsyncDisplayKit/Details/ASCollectionDataController.mm

```diff
@@ -237,7 +238,23 @@ - (void)_populateSupplementaryNodesOfKind:(NSString *)kind withSections:(NSIndex
 
   [sections enumerateRangesUsingBlock:^(NSRange range, BOOL * _Nonnull stop) {
     for (NSUInteger sec = range.location; sec < NSMaxRange(range); sec++) {
-      NSUInteger itemCount = [self.collectionDataSource dataController:self supplementaryNodesOfKind:kind inSection:sec];
+      // TODO: Unfortunately the data controller architecture stores supplementary elements in an array,
+      // instead of a dictionary. So we have to fake it by finding the highest index of the given
+      // supplementary element kind in this section, and make an array of supplementaries up to that index.
+      // Unfortunately this means we cannot support indexless supplementary elements.
+      NSUInteger itemCount = 0;
+      if (self.supportsDeclarativeData) {
+        for (NSNumber *indexObj in self.currentData.sectionsInternal[sec].supplementaryElements[kind]) {
+          NSInteger index = indexObj.integerValue;
+          if (index == NSNotFound) {
+            ASDisplayNodeFailAssert(@"Supplementary elements with no index are not currently supported. Support is planned for the near future.");
+          }
+          itemCount = MAX(itemCount, index + 1);
```

Comment:
I'm not really sure how to verify this logic is correct.  Seems like a real pain in the ass, will be great to tear apart the supplementary support and rebuild it next year :).  Do we have sufficient test cases to vet this?